rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+We’ve considered an alternative where
+[**StructuredTensor**](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, and **tf.Data** is the unified I/O
+abstraction.
+
+StructuredTensor is equally powerful as Arrow’s StructArray so it’s able to
+represent all our logical representations.
+
+Because StructuredTensor is a CompositeTensor, we could imagine that I/O +
+parsing of a logical data format in a physical storage format is handled by a
+specific tf.Data Dataset that yields StructuredTensors.
+
+We also need to be able to construct a beam source from such a Dataset (although
+there is a non-trivial gap).
+
+The advantages of this approach are:
+
+*   Less effort to integrate: adding a new format == adding a new tf.Dataset.
+*   Unified story across TF and beam.
+*   No third-party dependencies.
+    *   Note that we can still provide an Arrow -> StructuredTensor adapter to
+        achieve interoperability.
+
+What this alternative does not change / address:
+
+*   We still need the TensorAdapter API to convert from StructuredTensor to
+    other TF feedables, unless we are willing to only offer StructuredTensors to
+    our end-users (which might be the case eventually, but not likely to happen
+    soon). So a good portion of this proposal will remain mostly unchanged.
+*   We still cannot expose file-based or record-based interface to end-users in
+    the TF trainer. As that is a direct result of I/O + parsing being abstracted
+    out.
+
+The disadvantages of this approach are:
+
+*   This tightly couples TF with TFX components, in the following ways:
+
+    *   Components will need TF to read the data in.
+    *   Components that analyze the data (e.g. TFDV) will operate against
+        StructuredTensors, and the easiest way to to conduct certain
+        computations (for example, slicing a StructuredTensor or computing the
+        mean) with StructuredTensors is through TF ops and their python
+        bindings.
+
+    Such a coupling is not in our favor because:
+
+    *   Some TFX components functionally do not require TF to work. For example,
+        TFDV can analyze any data set. TFMA can analyze a model while treating
+        the model as a blackbox. In both cases, the ML library that trained the
+        model is irrelevant and TF should not be assumed. And coupling with TF
+        does not only introduce a heavy dependency, but also forces the user to
+        learn about TF if they need to implement a TFXIO for their data format.
+
+    *   Operations against StructuredTensors boils down to Python + TF ops. The
+        overhead of either is much higher than just calling an Arrow C++ API
+        that does the same operation.
+
+    *   The main extension point of TF is Ops which don't understand the nested
+        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
+        deals with StructuredTensors will be much complicated.
```

Comment:
I don't understand this point; can you elaborate? An example might help.

Reply:
I assume you agree that OPs are the main extension point of TF (i.e. if I want to do custom computation over TF (composite) tensors, I would want to implement a custom OP).

A TF OP that takes a StructuredTensor as input is more complex to implement due to:

1) there is no corresponding C++ data structure that can describe what StructuredTensor describes.
2) there is no C++ API to re-construct a CompositeTensor from flattened list of tensorflow::Tensors (which is the raw input interface to an OP).

And in general, implementing a TF OP involves much more boilerplate than implementing a custom function that takes Arrow entities.

Moreover, TF has more serious C++ ABI compatibility issues until Modular TF.

Maybe the above can be partially addressed by C++ composite tensor. But I don't know how exactly that's going to be implemented.

Reply:
No you don't need to implement an op to do computation over TF composite tensors; if the underlying tensors in the composite are stored on the CPU you can no-copy convert them into numpy arrays and no-copy convert resulting arrays back into tensors, so you can write your C++ code however you want to interact with them (in a post-eager-execution world).

You only need to implement a TF op if you want it to commute with tf.function, gradients, etc, which will likely not be the case for most TFX components.

Reply:
But imagine that I want to write a C++ function that takes the entire structure as input; due to lack of a C++ class for StructuredTensors, my C++ code can only see individual numpy arrays and have to recover the structure from those numpy arrays.

Reply:
I updated this section to compare StructuredTensor with RecordBatch per the requirements (the "requirements" section was also newly added). PTAL.

I also added a colab to elaborate what I meant here:
https://colab.sandbox.google.com/drive/1SCQs88J4Tc6HKk2AfpYvELaJDxa4h4IQ